Fixed Show Favorites on startup which was not working as it should

walkero-gr opened the pull request:
While I was investigating #100 I saw that "Show Favorites on StartUp" didn't work as it should, and the list was not shown by default. This PR has a fix for this.

walkero-gr commented on src/funcs.c:
```diff
@@ -596,10 +596,12 @@ void list_show_favorites(char* str)
 			const int length = strlen(helper);
 			for (int i = 0; i <= length - 1; i++) helper[i] = tolower(helper[i]);
 
-			if (item_games->favorite == 1 && item_games->hidden != 1 && strstr(helper, (char *)str))
```
When the start_with_favorites is 1, the **str** in this line is NULL (check line 800 which calls this method). So, this condition is always **false**.

walkero-gr commented on src/funcs.c:
```diff
@@ -596,10 +596,12 @@ void list_show_favorites(char* str)
 			{
-				DoMethod(app->LV_GamesList, MUIM_List_InsertSingle, item_games->title, MUIV_List_Insert_Sorted);
-				total_games++;
+				if (strstr(helper, (char *)str) || (str == NULL)) {
```
With this fix both showing favorites on startup and filtering from the input field work just fine.

coldacid replied:
Given this is using `strstr()` does this mean that the filtering is case-sensitive? Or do we lower-case the strings elsewhere first?

walkero-gr replied:
@coldacid Good question. I will check that today how it works. If it is case sensitive, we might need a new task for next version to change it.
Thanks

walkero-gr replied:
@coldacid Checked that filter and is case insensitive and can return entries even what you search is in the middle of the name.

walkero-gr commented:
Before we merge that, I am curious if "Show favorites on startup" works for you. This was added at f74acbcb971a02d026909f5bf8d60567c577c40b

walkero-gr commented:
I did some further investigation on this one, because I had some second thoughts and I have some findings. First of all I reverted the code, like
`if (item_games->favorite == 1 && item_games->hidden != 1 && strstr(helper, (char *)str))`

Before that I added the following code:
```
const char* pAnd = strstr(helper, (char *)str);
printf("%s - %s\n", pAnd, (char *)str);
```
Tested that on AmigaOS 3.1.4 with MUI 5.0 and I had the following output at the shell
![PR-102_mui5](https://user-images.githubusercontent.com/14014909/94200958-316c3e80-feb3-11ea-88d0-9b96d695b489.png)

Tested that on AmigaOS 3.1.4 with MUI 3.8 and I had the following output at the shell
![PR-102_mui3](https://user-images.githubusercontent.com/14014909/94201008-3fba5a80-feb3-11ea-90b5-7a48db6bc023.png)

For some reason the `strstr` gives bad results when the second parameter is NULL on MUI 5. I am trying to find a better way to do that. This is something that is used in different places in iGame as well, so a better solution is necessary.

I will test the strcasestr.c of @coldacid as seen at #96

walkero-gr commented:
Tested with strcasestr() and the results are the same. 
Tried also to build it with MUI 5 SDK but the results are the same.

coldacid commented:
Based on https://github.com/MrZammler/iGame/commit/2e3678e77766a63688d9c5397ed83905c6d8284b#diff-f70b30daa1ea19d477af3be6b35f2d8bR13-R15 `strcasestr()` should return a pointer to `haystack` if `needle` is an empty string, but I don't think that `NULL` itself is supported as a valid input for `needle` for either `strstr()` or `strcasestr()`. I'll take a look in my reference. It could be that we're facing bugs in both function's implementations, or it could be that we're misusing the functions in this situation.

coldacid commented:
@walkero-gr based on the last revision of C99, there is no defined behaviour for `strstr()` if either `haystack` or `needle` are null pointers. Therefore, we should be guarding the call ourselves if either argument is a null pointer instead of a pointer to a null string.

walkero-gr commented:
@coldacid @MrZammler Updated this PR. For the fix I created a new method str_exists(), which checks if the needle is NULL. If yes returns TRUE, because actually we do not need to filter the results. This code works on both MUI 3 and 5, and I believe is going to be a little bit faster when the needle is NULL, which is the case where the user set to start with favorites.

coldacid commented on src/funcs.c:
```diff
@@ -817,6 +836,8 @@ void filter_change()
 	get(app->STR_Filter, MUIA_String_Contents, &str);
 	DoMethod(app->LV_GenresList, MUIM_List_GetEntry, MUIV_List_GetEntry_Active, &str_gen);
 
+	printf("filter_change:%s~~~\n", str);
```
Should probably wrap this with `#ifdef DEBUG` / `#endif` or remove it.

walkero-gr replied:
Thanks for that. Removed.

coldacid commented:
I probably would have just turned the original `strstr(helper, (char *)str)` into `(str == NULL || strstr(helper, (char *)str)` rather than create a new function, to be honest. Like how your [original commit did](https://github.com/MrZammler/iGame/pull/102/files/50735607e4627a14dd0140874f40ec14c0507482#diff-6ea6b0ed6d5219f4a254df0382930532R601).

walkero-gr commented:
My consideration here is if on execution both parts would be checked, and strstr fills memory with trash

walkero-gr commented:
Got my answer at https://en.wikipedia.org/wiki/Short-circuit_evaluation
Reverting the code

walkero-gr commented:
@coldacid Reverted the code with the latest commit
